File: src/main/java/editor/database/card/SingleCard.java
package editor.database.card;

import java.util.Collections;
import java.util.Date;
import java.util.HashSet;
import java.util.List;
import java.util.Map;
import java.util.Set;
import java.util.TreeMap;

import javax.swing.text.BadLocationException;
import javax.swing.text.Style;
import javax.swing.text.StyleConstants;
import javax.swing.text.StyledDocument;

import editor.database.attributes.CombatStat;
import editor.database.attributes.Expansion;
import editor.database.attributes.Legality;
import editor.database.attributes.Loyalty;
import editor.database.attributes.ManaCost;
import editor.database.attributes.ManaType;
import editor.database.attributes.Rarity;
import editor.database.symbol.FunctionalSymbol;
import editor.database.symbol.Symbol;
import editor.gui.generic.ComponentUtils;
import editor.util.UnicodeSymbols;

/**
 * This class represents a single-faced Magic: the Gathering card.  All of the
 * methods that return sets of attributes will only contain one value.
 *
 * @author Alec Roelke
 */
public class SingleCard extends Card
{
    /**
     * Set containing all of the types of this SingleCard.
     */
    public final Set<String> allTypes;
    /**
     * This SingleCard's artist.
     */
    public final String artist;
    /**
     * This SingleCard's color identity.
     *
     * @see Card#colorIdentity()
     */
    private final List<ManaType> colorIdentity;
    /**
     * This SingleCard's colors.
     */
    public final List<ManaType> colors;
    /**
     * List of formats this SingleCard can be commander in.
     */
    public final List<String> commandFormats;
    /**
     * This SingleCard's flavor text.
     */
    public final String flavor;
    /**
     * This SingleCard's image name.
     */
    public final String imageName;
    /**
     * Whether or not this SingleCard is a land card.
     */
    public final boolean isLand;
    /**
     * Formats and legality for this Card.
     */
    private final Map<String, Legality> legality;
    /**
     * This SingleCard's loyalty.
     */
    public final Loyalty loyalty;
    /**
     * Mana cost of this SingleCard.
     */
    public final ManaCost mana;
    /**
     * multiverseid of this SingleCard.
     */
    public final int multiverseid;
    /**
     * Scryfall illustration ID of this SingleCard.
     */
    public final String scryfallid;
    /**
     * Name of this SingleCard.
     */
    public final String name;
    /**
     * This SingleCard's collector's number.
     */
    public final String number;
    /**
     * This SingleCard's power.
     *
     * @see Card#power()
     */
    public final CombatStat power;
    /**
     * The text actually printed on the card.
     */
    public final String printed;
    /**
     * The type line actually printed on the card.
     */
    public final String printedTypes;
    /**
     * This Card's rarity.
     */
    private final Rarity rarity;
    /**
     * Rulings for this Card.
     */
    private final Map<Date, List<String>> rulings;
    /**
     * This SingleCard's subtypes.
     */
    public final Set<String> subtypes;
    /**
     * This SingleCard's supertypes.
     */
    public final Set<String> supertypes;
    /**
     * This SingleCard's rules text.
     */
    public final String text;
    /**
     * This SingleCard's toughness.
     *
     * @see Card#toughness()
     */
    public final CombatStat toughness;
    /**
     * This SingleCard's type line.
     *
     * @see Card#typeLine()
     */
    public final String typeLine;
    /**
     * This SingleCard's types.
     */
    public final Set<String> types;

    /**
     * Create a new Card with a single face.
     *
     * @param layout the new Card's layout
     * @param name the new Card's name
     * @param mana the new Card's mana cost
     * @param colors the new Card's colors
     * @param colorIdentity the new Card's color identity
     * @param supertype the new Card's supertypes
     * @param type the new Card's types
     * @param subtype the new Card's subtypes
     * @param printedTypes the new Card's printed type line
     * @param rarity the new Card's rarity
     * @param set the Expansion the new Card belongs to
     * @param text the new Card's rules text
     * @param flavor the new Card's flavor text
     * @param printed the new Card's printed text
     * @param artist the new Card's artist
     * @param multiverseid the new Card's multiverse ID
     * @param scryfallid the new Card's Scryfall illustration ID
     * @param number the new Card's collector's number
     * @param power the new Card's power
     * @param toughness the new Card's toughness
     * @param loyalty the new Card's loyalty (0 or less for nonexistent)
     * @param legality the new Card's legality
     * @param command the new Card's command formats
     */
    public SingleCard(CardLayout layout,
                      String name,
                      ManaCost mana,
                      List<ManaType> colors,
                      List<ManaType> colorIdentity,
                      Set<String> supertype,
                      Set<String> type,
                      Set<String> subtype,
                      String printedTypes,
                      Rarity rarity,
                      Expansion set,
                      String text,
                      String flavor,
                      String printed,
                      String artist,
                      int multiverseid,
                      String scryfallid,
                      String number,
                      CombatStat power,
                      CombatStat toughness,
                      Loyalty loyalty,
                      TreeMap<Date, List<String>> rulings,
                      Map<String, Legality> legality,
                      List<String> command)
    {
        super(set, layout, 1);

        this.name = name;
        this.mana = mana;
        this.colors = Collections.unmodifiableList(colors);
        this.colorIdentity = Collections.unmodifiableList(colorIdentity);
        this.supertypes = Collections.unmodifiableSet(supertype);
        this.types = Collections.unmodifiableSet(type);
        this.subtypes = Collections.unmodifiableSet(subtype);
        this.printedTypes = printedTypes;
        this.text = text;
        this.flavor = flavor;
        this.printed = printed;
        this.artist = artist;
        this.number = number;
        this.multiverseid = multiverseid;
        this.scryfallid = scryfallid;
        this.power = power;
        this.toughness = toughness;
        this.loyalty = loyalty;
        this.imageName = name.toLowerCase();
        this.rarity = rarity;
        this.rulings = Collections.unmodifiableMap(rulings);
        this.legality = Collections.unmodifiableMap(legality);
        this.commandFormats = Collections.unmodifiableList(command);

        isLand = typeContains("land");

        // Create the type line for this Card
        StringBuilder str = new StringBuilder();
        if (supertypes.size() > 0)
            str.append(String.join(" ", supertypes)).append(" ");
        str.append(String.join(" ", types));
        if (subtypes.size() > 0)
            str.append(" " + UnicodeSymbols.EM_DASH + " ").append(String.join(" ", subtypes));
        typeLine = str.toString();

        var faceTypes = new HashSet<String>();
        faceTypes.addAll(supertypes);
        faceTypes.addAll(types);
        faceTypes.addAll(subtypes);
        allTypes = Collections.unmodifiableSet(faceTypes);
    }

    @Override
    public List<Set<String>> allTypes()
    {
        return Collections.singletonList(allTypes);
    }

    @Override
    public List<String> artist()
    {
        return Collections.singletonList(artist);
    }

    @Override
    public double manaValue()
    {
        return mana.manaValue();
    }

    @Override
    public double minManaValue()
    {
        return manaValue();
    }

    @Override
    public double maxManaValue()
    {
        return manaValue();
    }

    @Override
    public double avgManaValue()
    {
        return manaValue();
    }

    @Override
    public List<ManaType> colorIdentity()
    {
        return colorIdentity;
    }

    @Override
    public List<ManaType> colors()
    {
        return colors;
    }

    /**
     * {@inheritDoc}
     * This returns the same thing as {@link #colors()}.
     *
     * @throw IndexOutOfBoundsException if face is not equal to 0, since SingleCards only have
     * one face
     */
    @Override
    public List<ManaType> colors(int face) throws IndexOutOfBoundsException
    {
        if (face != 0)
            throw new IndexOutOfBoundsException("Single-faced cards only have one face");
        return colors;
    }

    @Override
    public List<String> commandFormats()
    {
        return commandFormats;
    }

    @Override
    public List<String> flavorText()
    {
        return Collections.singletonList(flavor);
    }

    @Override
    public List<String> imageNames()
    {
        return Collections.singletonList(imageName);
    }

    @Override
    public boolean isLand()
    {
        return isLand;
    }

    @Override
    public Map<String, Legality> legality()
    {
        return legality;
    }

    @Override
    public List<Loyalty> loyalty()
    {
        return Collections.singletonList(loyalty);
    }

    @Override
    public List<ManaCost> manaCost()
    {
        return Collections.singletonList(mana);
    }

    @Override
    public List<String> name()
    {
        return Collections.singletonList(name);
    }

    @Override
    public List<String> number()
    {
        return Collections.singletonList(number);
    }

    @Override
    public List<Integer> multiverseid()
    {
        return Collections.singletonList(multiverseid);
    }

    @Override
    public List<String> scryfallid()
    {
        return Collections.singletonList(scryfallid);
    }

    @Override
    public List<String> oracleText()
    {
        return Collections.singletonList(text);
    }

    @Override
    public List<CombatStat> power()
    {
        return Collections.singletonList(power);
    }

    @Override
    public List<String> printedText()
    {
        return Collections.singletonList(printed);
    }

    @Override
    public List<String> printedTypes()
    {
        return Collections.singletonList(printedTypes);
    }

    @Override
    public Rarity rarity()
    {
        return rarity;
    }

    @Override
    public Map<Date, List<String>> rulings()
    {
        return rulings;
    }

    @Override
    public Set<String> subtypes()
    {
        return subtypes;
    }

    @Override
    public Set<String> supertypes()
    {
        return supertypes;
    }

    @Override
    public List<CombatStat> toughness()
    {
        return Collections.singletonList(toughness);
    }

    @Override
    public List<String> typeLine()
    {
        return Collections.singletonList(typeLine);
    }

    @Override
    public Set<String> types()
    {
        return types;
    }

    @Override
    public void formatDocument(StyledDocument document, boolean printed)
    {
        Style textStyle = document.getStyle("text");
        Style reminderStyle = document.getStyle("reminder");
        Style chaosStyle = document.addStyle("CHAOS", null);
        StyleConstants.setIcon(chaosStyle, FunctionalSymbol.CHAOS.getIcon(ComponentUtils.TEXT_SIZE));
        try
        {
            document.insertString(document.getLength(), name + " ", textStyle);
            if (!mana.isEmpty())
            {
                for (Symbol symbol : mana)
                {
                    Style style = document.addStyle(symbol.toString(), null);
                    StyleConstants.setIcon(style, symbol.getIcon(ComponentUtils.TEXT_SIZE));
                    document.insertString(document.getLength(), symbol.toString(), style);
                }
                document.insertString(document.getLength(), " ", textStyle);
            }
            if (mana.manaValue() == (int)mana.manaValue())
                document.insertString(document.getLength(), "(" + (int)mana.manaValue() + ")\n", textStyle);
            else
                document.insertString(document.getLength(), "(" + mana.manaValue() + ")\n", textStyle);
            if (!mana.colors().equals(colors))
            {
                for (ManaType color : colors)
                {
                    Style indicatorStyle = document.addStyle("indicator", document.getStyle("text"));
                    StyleConstants.setForeground(indicatorStyle, color.color);
                    document.insertString(document.getLength(), String.valueOf(UnicodeSymbols.BULLET), indicatorStyle);
                }
                if (!colors().isEmpty())
                    document.insertString(document.getLength(), " ", textStyle);
            }
            if (printed)
                document.insertString(document.getLength(), printedTypes + '\n', textStyle);
            else
                document.insertString(document.getLength(), typeLine + '\n', textStyle);
            document.insertString(document.getLength(), expansion().name() + ' ' + rarity() + '\n', textStyle);

            String abilities = printed ? this.printed : text;
            if (!abilities.isEmpty())
            {
                int start = 0;
                Style style = textStyle;
                for (int i = 0; i < abilities.length(); i++)
                {
                    if (i == 0 || abilities.charAt(i) == '\n')
                    {
                        int nextLine = abilities.substring(i + 1).indexOf('\n');
                        int dash = (nextLine > -1 ? abilities.substring(i, nextLine + i) : abilities.substring(i)).indexOf(UnicodeSymbols.EM_DASH);
                        if (dash > -1)
                        {
                            dash += i;
                            if (i > 0)
                                document.insertString(document.getLength(), abilities.substring(start, i), style);
                            start = i;
                            // Assume that the dash used for ability words is surrounded by spaces, but not
                            // for keywords with cost parameters when the cost is nonmana cost
                            if (abilities.charAt(dash - 1) == ' ' && abilities.charAt(dash + 1) == ' ')
                                style = reminderStyle;
                        }
                    }
                    switch (abilities.charAt(i))
                    {
                    case '{':
                        document.insertString(document.getLength(), abilities.substring(start, i), style);
                        start = i + 1;
                        break;
                    case '}':
                        var symbol = Symbol.tryParseSymbol(abilities.substring(start, i));
                        if (symbol.isEmpty())
                        {
                            System.err.println("Unexpected symbol {" + abilities.substring(start, i) + "} in oracle text for " + unifiedName() + ".");
                            document.insertString(document.getLength(), abilities.substring(start, i), textStyle);
                        }
                        else
                        {
                            Style symbolStyle = document.addStyle(symbol.get().toString(), null);
                            StyleConstants.setIcon(symbolStyle, symbol.get().getIcon(ComponentUtils.TEXT_SIZE));
                            document.insertString(document.getLength(), symbol.get().toString(), symbolStyle);
                        }
                        start = i + 1;
                        break;
                    case '(':
                        document.insertString(document.getLength(), abilities.substring(start, i), style);
                        style = reminderStyle;
                        start = i;
                        break;
                    case ')':
                        document.insertString(document.getLength(), abilities.substring(start, i + 1), style);
                        style = textStyle;
                        start = i + 1;
                        break;
                    case 'C':
                        if (i < abilities.length() - 5 && abilities.substring(i, i + 5).equals("CHAOS"))
                        {
                            document.insertString(document.getLength(), abilities.substring(start, i), style);
                            document.insertString(document.getLength(), "CHAOS", chaosStyle);
                            start = i += 5;
                        }
                        break;
                    case UnicodeSymbols.EM_DASH:
                        document.insertString(document.getLength(), abilities.substring(start, i), style);
                        style = textStyle;
                        start = i;
                        break;
                    default:
                        break;
                    }
                    if (i == abilities.length() - 1 && abilities.charAt(i) != '}' && abilities.charAt(i) != ')')
                        document.insertString(document.getLength(), abilities.substring(start, i + 1), style);
                }
                document.insertString(document.getLength(), "\n", textStyle);
            }
            if (!flavor.isEmpty())
            {
                int start = 0;
                for (int i = 0; i < flavor.length(); i++)
                {
                    switch (flavor.charAt(i))
                    {
                    case '{':
                        document.insertString(document.getLength(), flavor.substring(start, i), reminderStyle);
                        start = i + 1;
                        break;
                    case '}':
                        var symbol = Symbol.tryParseSymbol(flavor.substring(start, i));
                        if (symbol.isEmpty())
                        {
                            System.err.println("Unexpected symbol {" + flavor.substring(start, i) + "} in flavor text for " + unifiedName() + ".");
                            document.insertString(document.getLength(), flavor.substring(start, i), reminderStyle);
                        }
                        else
                        {
                            Style symbolStyle = document.addStyle(symbol.get().toString(), null);
                            StyleConstants.setIcon(symbolStyle, symbol.get().getIcon(ComponentUtils.TEXT_SIZE));
                            document.insertString(document.getLength(), " ", symbolStyle);
                        }
                        start = i + 1;
                        break;
                    default:
                        break;
                    }
                    if (i == flavor.length() - 1 && flavor.charAt(i) != '}')
                        document.insertString(document.getLength(), flavor.substring(start, i + 1), reminderStyle);
                }
                document.insertString(document.getLength(), "\n", reminderStyle);
            }

            if (power.exists() && toughness.exists())
                document.insertString(document.getLength(), power + "/" + toughness + "\n", textStyle);
            else if (loyalty.exists())
                document.insertString(document.getLength(), loyalty + "\n", textStyle);

            document.insertString(document.getLength(), artist + " " + number + "/" + expansion().count(), textStyle);
        }
        catch (BadLocationException e)
        {
            e.printStackTrace();
        }
    }

    @Override
    public void formatDocument(StyledDocument document, boolean printed, int face)
    {
        formatDocument(document, printed);
    }
}

File: src/main/scala/editor/database/card/SingleCard.scala
package editor.database.card

import editor.database.attributes.CombatStat
import editor.database.attributes.Expansion
import editor.database.attributes.Legality
import editor.database.attributes.CounterStat
import editor.database.attributes.ManaCost
import editor.database.attributes.ManaType
import editor.database.attributes.Rarity
import editor.database.attributes.TypeLine
import editor.database.symbol.FunctionalSymbol
import editor.database.symbol.Symbol
import editor.gui.generic.ComponentUtils
import editor.unicode.{_, given}

import java.util.Date
import javax.swing.text.BadLocationException
import javax.swing.text.Style
import javax.swing.text.StyleConstants
import javax.swing.text.StyledDocument
import scala.collection.immutable.ListSet
import scala.collection.immutable.TreeMap
import scala.util.matching._
import editor.database.symbol.ColorSymbol
import editor.database.symbol.ManaSymbolInstances

/**
 * A single-faced [[Card]], or a single face of a [[MultiCard]].
 * 
 * @constructor create a new single-faced card.
 * @param layout layout of the card does not have to be a single-faced layout, but multi-faced ones should later
 * be joined together using the corresponding class
 * @param name name of the card
 * @param manaCost manaCost cost of the card
 * @param colors colors of the card does not have to correspond with the colors of its manaCost cost (but usually does)
 * @param colorIdentity color identity of the card
 * @param superts supertype set of the card (should be sorted in order of appearance)
 * @param cardts type set of the card (should be sorted in order of appearance)
 * @param subts subtype set of the card (should be sorted in order of appearance)
 * @param printedTypes printed type line of the card using its original wording
 * @param rarity rarity of the card
 * @param set expansion to which the card belongs
 * @param oracleText Oracle text of the card
 * @param flavorText flavor text of the card
 * @param printedText printed text of the card using its original wording
 * @param artist artist who illustrated the card
 * @param multiverseid multiverseid of the card (unique ID used in Gatherer to identify cards)
 * @param scryfallid Scryfall ID of the card
 * @param number collector number of the card
 * @param power power of the card, if it's a creature
 * @param toughness toughness of the card, if it's a creature
 * @param loyalty loyalty of the card, if it's a planeswalker
 * @param defense defense of the card, if it's a battle
 * @param rulings clarifications on how the card works and when they were made
 * @param legality which formats the card is legal (or restricted) in
 * @param commandFormats formats in which the card can be commander
 */
case class SingleCard(
  override val layout: CardLayout,
  name: String,
  manaCost: ManaCost,
  colors: Set[ManaType],
  colorIdentity: Set[ManaType],
  private val superts: ListSet[String],
  private val cardts: ListSet[String],
  private val subts: ListSet[String],
  printedTypes: String,
  rarity: Rarity,
  set: Expansion,
  oracleText: String,
  flavorText: String,
  printedText: String,
  artist: String,
  multiverseid: Int,
  scryfallid: String,
  number: String,
  power: Option[CombatStat],
  toughness: Option[CombatStat],
  loyalty: Option[CounterStat],
  defense: Option[CounterStat],
  rulings: TreeMap[Date, Seq[String]],
  legality: Map[String, Legality],
  commandFormats: Seq[String]
) extends Card(set, layout) {
  override def faces = Seq(this)

  override def manaValue = manaCost.manaValue
  override def minManaValue = manaCost.manaValue
  override def maxManaValue = manaCost.manaValue
  override def avgManaValue = manaCost.manaValue
  override lazy val typeLine = TypeLine(cardts, subts, superts)
  override lazy val isLand = types.exists(_.equalsIgnoreCase("land"))
  override lazy val imageNames = Seq(name.toLowerCase)

  private val specificPattern = raw"adds? (?:an amount of |\w+ |an additional (?:amount of |\w+ )?)?((?:(?:, |,? or )?${ManaCost.Pattern.regex})+)".r
  private val anyPattern = raw"(?s).*add (?:\w+|an amount of) mana (?:of any|of any one|in any combination of) (type|color).*".r

  override lazy val produces = normalizedOracle.head match {
    case anyPattern("color") => ManaType.colors.toSet
    case anyPattern("type")  => ManaType.values.toSet
    case _ =>
      val matches = specificPattern.findAllIn(normalizedOracle.head).matchData
      if (matches.isEmpty) {
        Set.empty
      } else {
        matches.flatMap((m) => ManaCost.parse(m.group(1).replaceAll("( |,|or)", ""))).map(_.intensity.keys).flatten.toSet
      } ++ Map(
        "plains" -> ManaType.White,
        "island" -> ManaType.Blue,
        "swamp" -> ManaType.Black,
        "mountain" -> ManaType.Red,
        "forest" -> ManaType.Green
      ).collect{ case (word, color) if subtypes.exists(_.equalsIgnoreCase(word)) => color }.toSet
  }

  override def formatDocument(document: StyledDocument, printed: Boolean) = {
    val textStyle = document.getStyle("text")
    val reminderStyle = document.getStyle("reminder")
    val chaosStyle = document.addStyle("CHAOS", null)
    StyleConstants.setIcon(chaosStyle, FunctionalSymbol.Chaos.scaled(ComponentUtils.TextSize))
    try {
      document.insertString(document.getLength(), s"${name} ", textStyle)
      if (!manaCost.isEmpty) {
        for (symbol <- manaCost) {
          val style = document.addStyle(symbol.toString, null)
          StyleConstants.setIcon(style, symbol.scaled(ComponentUtils.TextSize))
          document.insertString(document.getLength, symbol.toString, style)
        }
        document.insertString(document.getLength, " ", textStyle)
      }
      if (manaCost.manaValue == manaCost.manaValue.toInt)
        document.insertString(document.getLength(), s"(${manaCost.manaValue.toInt})\n", textStyle)
      else
        document.insertString(document.getLength(), s"(${manaCost.manaValue})\n", textStyle)
      if (manaCost.colors != colors) {
        for (color <- colors) {
          val indicatorStyle = document.addStyle("indicator", document.getStyle("text"))
          color.color.foreach(StyleConstants.setForeground(indicatorStyle, _))
          document.insertString(document.getLength, Bullet, indicatorStyle)
        }
        if (!colors.isEmpty)
          document.insertString(document.getLength, " ", textStyle)
      }
      if (printed)
        document.insertString(document.getLength, s"$printedTypes\n", textStyle)
      else
        document.insertString(document.getLength, s"${typeLine}\n", textStyle)
      document.insertString(document.getLength, s"${expansion.name} $rarity\n", textStyle)

      val abilities = if (printed) printedText else oracleText
      if (!abilities.isEmpty) {
        var i = 0
        var start = 0
        var style = textStyle
        while (i < abilities.size) {
          if (i == 0 || abilities(i) == '\n') {
            val nextLine = abilities.substring(i + 1).indexOf('\n')
            var dash = (if (nextLine > -1) abilities.substring(i, nextLine + i) else abilities.substring(i)).indexOf(EmDash)
            if (dash > -1) {
              dash += i
              if (i > 0)
                document.insertString(document.getLength, abilities.substring(start, i), style)
              start = i
              // Assume that the dash used for ability words is surrounded by spaces, but not
              // for keywords with cost parameters when the cost is nonmana cost
              if (abilities(dash - 1) == ' ' && abilities(dash + 1) == ' ')
                style = reminderStyle
            }
          }
          abilities(i) match {
            case '{' =>
              document.insertString(document.getLength, abilities.substring(start, i), style)
              start = i + 1
            case '}' =>
              val symbol = Symbol.parse(abilities.substring(start, i))
              if (symbol.isEmpty) {
                System.err.println(s"Unexpected symbol {${abilities.substring(start, i)}} in oracle text for $name.")
                document.insertString(document.getLength, abilities.substring(start, i), textStyle)
              } else {
                val symbolStyle = document.addStyle(symbol.get.toString, null)
                StyleConstants.setIcon(symbolStyle, symbol.get.scaled(ComponentUtils.TextSize))
                document.insertString(document.getLength, symbol.get.toString, symbolStyle)
              }
              start = i + 1
            case '(' =>
              document.insertString(document.getLength, abilities.substring(start, i), style)
              style = reminderStyle
              start = i
            case ')' =>
              document.insertString(document.getLength, abilities.substring(start, i + 1), style)
              style = textStyle
              start = i + 1
            case 'C' =>
              if (i < abilities.size - 5 && abilities.substring(i, i + 5) == "CHAOS") {
                document.insertString(document.getLength, abilities.substring(start, i), style)
                document.insertString(document.getLength, "CHAOS", chaosStyle)
                i += 5
                start = i
              }
              if (abilities(i) == '}')
                start += 1
            case EmDash.toChar =>
              document.insertString(document.getLength, abilities.substring(start, i), style)
              style = textStyle
              start = i
            case _ =>
          }
          if (i == abilities.size - 1 && abilities(i) != '}' && abilities(i) != ')')
            document.insertString(document.getLength, abilities.substring(start, i + 1), style)
          i += 1
        }
        document.insertString(document.getLength, "\n", textStyle)
      }
      if (!flavorText.isEmpty) {
        var i = 0
        var start = 0
        while (i < flavorText.size) {
          flavorText(i) match {
            case '{' =>
              document.insertString(document.getLength, flavorText.substring(start, i), reminderStyle)
              start = i + 1
            case '}' =>
              val symbol = Symbol.parse(flavorText.substring(start, i))
              if (symbol.isEmpty) {
                System.err.println(s"Unexpected symbol {${flavorText.substring(start, i)}} in flavor text for $name.")
                document.insertString(document.getLength, flavorText.substring(start, i), reminderStyle)
              } else {
                val symbolStyle = document.addStyle(symbol.get.toString, null)
                StyleConstants.setIcon(symbolStyle, symbol.get.scaled(ComponentUtils.TextSize))
                document.insertString(document.getLength, " ", symbolStyle)
              }
              start = i + 1
            case _ =>
          }
          if (i == flavorText.length - 1 && flavorText(i) != '}')
            document.insertString(document.getLength, flavorText.substring(start, i + 1), reminderStyle)
          i += 1
        }
        document.insertString(document.getLength, "\n", reminderStyle)
      }

      (power, toughness, loyalty, defense) match {
        case (Some(p), Some(t), None, None) => document.insertString(document.getLength, s"$p/$t\n", textStyle)
        case (None, None, Some(l), None)    => document.insertString(document.getLength, s"[$l]\n", textStyle)
        case (None, None, None, Some(d))    => document.insertString(document.getLength, s"{$d}\n", textStyle)
        case (None, None, None, None)       => // don't add anything
        case _ => throw IllegalArgumentException(s"Illegal combination of power/toughness, loyalty, defense: $power/$toughness, $loyalty, $defense")
      }

      document.insertString(document.getLength, s"$artist $number/${expansion.count}", textStyle)
    } catch case e: BadLocationException => e.printStackTrace()
  }

  override def formatDocument(document: StyledDocument, printed: Boolean, face: Int) = formatDocument(document, printed)
}